Add VF2PostLayout to the C Transpiler pipeline

mtreinish opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adds the vf2 post layout transpiler pass to the rust
transpile() function used by qk_transpile() in C. This was not included
originally because there wasn't a path to running vf2postlayout from
rust before but since this has been added now we can run the pass as
part of the C transpilation.

One detail that was added as part of this the function for filling out a
post layout with idle physical qubits was missing from the vf2 post
implementation in rust since for the Python pass it was done Python side
still. This was added as a standalone function for rust since we need it
in rust, but the Python implementation still makes sense for the python
side.

### Details and comments

This PR is based on top of: https://github.com/Qiskit/qiskit/pull/14863 and will need to be rebased once that merges. In the meantime you can view the contents of just this PR by looking at the HEAD commit: https://github.com/Qiskit/qiskit/commit/4eda52604b4ee8acb8114a68b9985f397b3a0dc3

mtreinish committed:
Add VF2PostLayout to the C Transpiler pipeline

This commit adds the vf2 post layout transpiler pass to the rust
transpile() function used by qk_transpile() in C. This was not included
originally because there wasn't a path to running vf2postlayout from
rust before but since this has been added now we can run the pass as
part of the C transpilation.

One detail that was added as part of this the function for filling out a
post layout with idle physical qubits was missing from the vf2 post
implementation in rust since for the Python pass it was done Python side
still. This was added as a standalone function for rust since we need it
in rust, but the Python implementation still makes sense for the python
side.

jakelishman commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -307,6 +307,8 @@ pub fn optimization_stage(
     commutation_checker: &mut CommutationChecker,
     equivalence_library: &mut EquivalenceLibrary,
+    transpile_layout: &mut TranspileLayout,
+    vf2_config: &vf2::Vf2PassConfiguration,
```
This is probably indicating that we should think about encapsulating structures - imo we should be paying more attention to `clippy::too_many_arguments` except for `pyfunction` (where they appear only because of kwargs).

mtreinish replied:
I can add a struct for the arguments. The only caller is here for the full transpile() and in the C API for the standalone function (which is in other PRs). The c api interface is simpler (since we don't expose the commutation checker, equiv library, vf2 config) so I really was just thinking this as internal. But using a struct doesn't hurt.

mtreinish replied:
I ended up removing the extra arg because the vf2 config needed to be different between pre and post so this is kind of a moot point now. I could still add a struct if you want, but it's not that many args now.

jakelishman commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -124,25 +124,9 @@ pub fn layout_stage(
-    };
-
+    vf2_config: &vf2::Vf2PassConfiguration,
+) -> Result<bool> {
```
We should at least document what the return value means.  Might be worth introducing a `LayoutSource` enum like `VF2`, `Sabre` to make it more self-documenting?

mtreinish replied:
Done in https://github.com/Qiskit/qiskit/pull/15151/commits/d79476252e6d139b0931cf355b824c399b260495

jakelishman commented on crates/transpiler/src/passes/apply_layout.rs:
```diff
@@ -117,6 +118,24 @@ pub fn update_layout(
+        }
+    }
+}
+
```
Ah nuts, I think this is a remnant bug in the Rust-space VF2Post - I'm fairly sure it's not _supposed_ to output a partial layout.  We fixed it in `VF2Layout` in Python space while the patch series was in flight, and I think I intended to update it in the Rust-space ones as well but forgot.

Assuming I'm remembering correctly, I think it's worth fixing this in the new `VF2Post` rather than patching it on here - it's just a footgun if there's a function you _always_ have to call after `vf2_layout_exact` to make it safe.

mtreinish replied:
I was thinking it depended on your use more than being always required. Like if you wanted to find the best active qubit mapping or something. But I can move it to be part of vf2post if you think that's better, I'll look at what you're doing in vf2layout and add that instead.

mtreinish replied:
I've removed this function in https://github.com/Qiskit/qiskit/pull/15151/commits/27727b78b85c6b40b5e9617fdce9ef30c371c47d it seems to work without it now. I think I just was using things incorrectly before or saw a need when there was a bug in an earlier version of the PR series and never tried it again after.

jakelishman commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -385,6 +387,14 @@ pub fn optimization_stage(
+            update_layout(dag, transpile_layout, |x| {
+                Qubit(layout[&VirtualQubit(x.0)].0)
+            });
+        }
```
the `vf2_config` struct used here is wrong; it sets `score_initial_layout: false`, which we don't usually want with `vf2_layout_pass_exact`, since we exact the initial layout to already be perfect and to have an associated score we can use to prune.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/15151/commits/8eecfeb98b72622a8abb8d3e1c21a8e09a8f61cb

mtreinish committed:
Use enum for layout source instead of boolean

mtreinish committed:
Remove reused vf2_config

mtreinish committed:
Merge remote-tracking branch 'origin/main' into vf2post-in-c-transpile

mtreinish committed:
Fix lint

mtreinish committed:
Remove allocate_idle_qubits

After rebasing this and after simplifying the code this function is no
longer needed so we can remove it and rely on the pass code to return a
full layout.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into pr/mtreinish/15151

raynelfss committed:
Fix: Add provisional arguments to C standalone stages
- Since the last time Matthew worked on this PR. All the standalone C API transpiler stages have been added, and they rely on the rust-side stages to function. Since this PR changes the arguments for certain stages. I have added provisional default values that may be removed in future iterations. These are based on the default settings of the transpiler.

raynelfss committed:
Fix: Add condition when running vf2
- An error in a previous implementation had the VF2 pass run only on optimization level 3 but had enabled configurations for 1 and 2. The confguration and run steps of the pass were moved outisde of this condition.
- VF2 is not used whenever we operate in optimization level 0, so another condition was added to only run the pass if the level is other than zero.

raynelfss committed:
Fix: Include idle qubits in vf2 mapping.
- The rust version of VF2Post forgets about the idle bits in the mapping that is returned, resulting in a possible panic in the next stages if the bit instance is not found. The following commit adds a way to keep track of the idling physical qubits and map the virtual ones onto them.

Cryoris commented on crates/cext/src/transpiler/transpile_function.rs:
```diff
@@ -388,6 +389,14 @@ pub unsafe extern "C" fn qk_transpile_stage_optimization(
         // and aligned pointer.
         unsafe { const_ptr_as_ref(options) }
     };
+    // TODO: Remove this.
```
Why should this be removed if the layout is required in the optimization stage?

raynelfss replied:
Because this should be passed as an argument instead of having it be a default value. This is just a reminder to remove that part.

**Update:** Just to clarify it needs to be an argument to the `optimization_stage` function call...

Cryoris replied:
I might be missing something, but it doesn't look like there's still a `layout` as argument on current main?

raynelfss replied:
No, it was added by this PR. Not quite sure why though. I'm assuming we use update our layout after running VF2Post during the last part of optimization.

Cryoris replied:
Looking at this in more detail, it doesn't seem valid to just pass a placeholder here. The optimization stage might actually change the layout and this change is lost right now. To keep this information we would actually need to require the transpile layout as (mutable) input

mtreinish replied:
This is why there was a todo to remove this, this needs a mutable reference to the transpile layout  on the function signature for this stage function. It should take in the transpile layout built up in the earlier atages. We should do this in 2.3 because we're exposing this function to c now and we want that in the signature in c to have the correct function signature for compatibility. You can see what the routing stage signature looks like in c.

This was a to-do when I wrote the pr because it was all in flux as we were still splitting out the stages as separate callables at the same time as the final vf2postlayout in rust prs were merging.

Cryoris commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -258,6 +273,32 @@ pub fn routing_stage(
+        let vf2_config = match optimization_level {
+            OptimizationLevel::Level0 => vf2::Vf2PassConfiguration::default_abstract(), // Not used.
+            OptimizationLevel::Level1 | OptimizationLevel::Level2 => vf2::Vf2PassConfiguration {
+                call_limit: (Some(5_000_000), Some(10_000)),
```
Are these hardcoded values what we also use  in Python? (Looking for their justification 😛 )

raynelfss replied:
I actually don't know exactly why Matt chose those values. Since the call limit here is a VF2 thing from rustworkx, I assume he chose this based on an assumption but I quite honestly don't have the answer here.

Cryoris replied:
Yeah they seem to be the same ones as is the preset pass manager in Python, e.g. https://github.com/Qiskit/qiskit/blob/bc2d374c4bf8748611530b6d769288dcf309c651/qiskit/transpiler/preset_passmanagers/builtin_plugins.py#L785

Cryoris commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -753,7 +824,13 @@ mod tests {
+                &target,
+                opt_level.into(),
+                Some(1.0),
+                Some(2_025_120_142),
```
I'm curious if this change is because we overused 42 as seed or if it has a more sinister reason (like the test failing with 42)?

raynelfss replied:
I'm not actually sure here either. If anything he might have just encoding a date? `2025_12_01_42` do as you may with that 🙃

mtreinish replied:
This was a leftover from testing. I didn't mean to commit it, I was checking various seeds to see if it all worked correctly with vf2post being run. 42 still works fine too.

Cryoris commented on crates/transpiler/src/transpiler.rs:
Shouldn't the layout source here also be "trivial"?

raynelfss replied:
Yeah, I'll admit I'm not too familiar with the layout stage. Matthew did put that comment there, do you think we would benefit from having it be trivial and not default to Sabre?

Cryoris replied:
Ok so it seems it doesn't matter because all functions that take `layout_source` as input don't actually do anything with it at optimization level 0. This seems like a potential sharp bit -- I would prefer still setting it to `Trivial` or at least put a big comment stating that it's not used so we don't bother setting it.

mtreinish replied:
Right level0 doesn't matter here because we never will run vf2post with level 0, level 0 always only uses trivial layout. You can change this to set trivial here but it doesn't really matter, the value is only used to control whether we run vf2post in the routing stage.

If you look back in the pr branch history you'll see this was a boolean whether we need to run vf2post or not and I changed it to the enum based on review feedback. But I forgot to update this because level 0 doesn't matter for tracking.

Cryoris commented:
Either I'm not understand the flow correctly or there's some information not being passed properly: I would expect the layouting stage to return the layout + source and then both these being consumed by routing, optimization and following passes. Patching default values seems like information might get lost. I would rather defer this to 2.4 and ensure we're getting this flow right before introducing potential bugs

mtreinish commented:
> Either I'm not understand the flow correctly or there's some information not being passed properly: I would expect the layouting stage to return the layout + source and then both these being consumed by routing, optimization and following passes. Patching default values seems like information might get lost. I would rather defer this to 2.4 and ensure we're getting this flow right before introducing potential bugs

The missing piece is the todo in the optimization stage which needs to be replaced with a mutable reference to the transpile layout (which is why it was a todo). The layout source isn't metadata needed after routing though it's just used to control whether we need to run vf2post in routing or not. The reason to do this in 2.3 though is this requires changes to the signatures on the stage functions in the c API which haven't been released yet. After 2.3.0rc1 this requires a breaking change to update the signatures to account for the layout in the optimization stage and whether we need to run vf2post in routing.

jakelishman commented:
I'm with Julien that I think we're better deferring the PR at this point - the RC release is supposed to be today, and we're already late on CF (and then `Target` as a knock-on effect from me).

The signatures in these functions won't be the final ones anyway (since we're pretty sure we're going to need to pass a more complete "state" object), so I don't think it's the end of the world if they're imperfect in this release.  It would technically a breaking change, but it's within our policy for the C API and unlike things to do with `Target` or the circuits, we're not really expecting any major downstream projects to be dependent on these stage functions yet anyway, so I don't think the impact would be large.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into vf2post-in-c-transpile

mtreinish committed:
Add missing layout argument

This commit adds the missing layout argument to the optimization stage.
Now that we're running vf2postlayout in the transpile optimization stage
function we need to take in a layout and update it based on the new
layout the pass generates. This is a backwards incompatible API change
as this new argument is required for callers. This is documented as such
in the release notes.

mtreinish committed:
Appease clippy

mtreinish committed:
Revert "Fix: Add condition when running vf2"

The previous implementation was correct, VF2PostLayout only runs in the
optimization stage at level 3. As this pass is very expensive and we
already run it in the routing stage and unlikely to find a better
solution this is something we only want to run in the highest
optimization level. This reverts commit
50bfa3f465b3ae1fb643d6c410c06504e81d568b which incorrectly changed the
PR.

mtreinish committed:
Set LayoutSource for routing stage to sabre

The LayoutSource enum being set to Trivial prevents VF2PostLayout from
running in the routing stage. This eliminates the purpose of adding the
pass to the stage. In practice the standalone functions in C can't know
whether to run or not because the previous stage is decoupled from it.
So it's better to err on the side of caution and always run the pass
even if we've already found the same solution.

mtreinish committed:
Include newly required layout argument in tests

mtreinish committed:
Merge branch 'main' into vf2post-in-c-transpile

mtreinish committed:
Merge branch 'main' into vf2post-in-c-transpile

raynelfss commented on test/c/test_transpiler.c:
```diff
@@ -416,7 +416,13 @@ int test_optimization_stage_empty(void) {
     QkQuantumRegister *qr = qk_quantum_register_new(2048, "qr");
     qk_dag_add_quantum_register(dag, qr);
-    int compile_result = qk_transpile_stage_optimization(dag, target, NULL, NULL);
+    uint32_t *layout_mapping = malloc(sizeof(uint32_t) * 2048);
```
Do we need to free the `layout_mapping` array here or does it get consumed by the Rust-side?

mtreinish replied:
Good call I need to free the allocation here I forgot about this.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/15151/commits/fd54df29e47a58aa2cc4371b9e55ec3bedca46bb

raynelfss commented on releasenotes/notes/vf2post-in-c-transpile-0228cfd71fcbf18b.yaml:
```diff
@@ -0,0 +1,17 @@
+    The :c:func:`qk_transpile` function now will run the :class:`.VF2PostLayout` transpiler pass as part
+    of optimization levels 1, 2, and 3. The pass is configured to run in the same way as the Python
+    preset pass managers run in :py:func:`.generate_preset_pass_manager`. You can refer to the Python
+    :mod`~.transpiler` documentation for more details on the configuration.
```
```suggestion
    :mod:`~.transpiler` documentation for more details on the configuration.
```

raynelfss reviewed (commented):
This looks about ready to merge. Just a couple observations.

mtreinish committed:
Update releasenotes/notes/vf2post-in-c-transpile-0228cfd71fcbf18b.yaml

Co-authored-by: Raynel Sanchez <87539502+raynelfss@users.noreply.github.com>

mtreinish committed:
Add missing free from test

raynelfss reviewed (approved):
LGTM!
